message/infrastructure/googlecloud/subscriber.go

```diff
@@ -0,0 +1,221 @@
+	"github.com/pkg/errors"
+)
+
+var (
+	ErrSubscriberClosed = errors.New("subscriber is closed")
+)
+
+type subscriber struct {
+	closing chan struct{}
+	closed  bool
+
+	allSubscriptionsWaitGroup sync.WaitGroup
+	activeSubscriptions       map[string]struct{}
+	activeSubscriptionsLock   sync.RWMutex
+
+	client *pubsub.Client
+	config SubscriberConfig
+
+	unmarshaler Unmarshaler
+	logger      watermill.LoggerAdapter
+}
+
+type SubscriberConfig struct {
+	SubscriptionName            string
+	ProjectID                   string
+	CreateSubscriptionIfMissing bool
+
+	ConsumersCount int
+
+	SubscriptionConfig pubsub.SubscriptionConfig
+	ClientOptions      []option.ClientOption
+	Unmarshaler        Unmarshaler
+}
+
+func (c *SubscriberConfig) setDefaults() {
+	if c.ConsumersCount == 0 {
+		c.ConsumersCount = runtime.NumCPU()
+	}
+
+	if c.Unmarshaler == nil {
+		c.Unmarshaler = DefaultUnmarshaler{}
+	}
+}
+
+func (c SubscriberConfig) Validate() error {
+	if c.ConsumersCount <= 0 {
+		return errors.Errorf("ConsumersCount must be greater than 0, have %d", c.ConsumersCount)
+	}
+	if c.SubscriptionName == "" {
+		return errors.New("SubscriptionName must be set")
+	}
+
+	if c.Unmarshaler == nil {
+		return errors.New("empty googlecloud message unmarshaler")
+	}
+
+	return nil
+}
+
+func NewSubscriber(
+	ctx context.Context,
+	config SubscriberConfig,
+	logger watermill.LoggerAdapter,
+) (message.Subscriber, error) {
+	config.setDefaults()
+	if err := config.Validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid config")
+	}
+
+	client, err := pubsub.NewClient(ctx, config.ProjectID, config.ClientOptions...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &subscriber{
+		closing: make(chan struct{}, 1),
+		closed:  false,
+
+		allSubscriptionsWaitGroup: sync.WaitGroup{},
+		activeSubscriptions:       map[string]struct{}{},
+		activeSubscriptionsLock:   sync.RWMutex{},
+
+		client: client,
+		config: config,
+
+		unmarshaler: config.Unmarshaler,
+		logger:      logger,
+	}, nil
+}
+
+func (s *subscriber) Subscribe(topic string) (chan *message.Message, error) {
+	// todo: pass root ctx from somewhere?
+	ctx, cancel := context.WithCancel(context.Background())
+
+	if s.closed {
+		return nil, ErrSubscriberClosed
+	}
+
+	logFields := watermill.LogFields{
+		"provider":          ProviderName,
+		"topic":             topic,
+		"consumers_count":   s.config.ConsumersCount,
+		"subscription_name": s.config.SubscriptionName,
+	}
+	s.logger.Info("Subscribing to Google Cloud PubSub topic", logFields)
+
+	output := make(chan *message.Message, 0)
+
+	s.allSubscriptionsWaitGroup.Add(1)
+
+	sub, err := s.subscription(ctx, topic)
+	if err != nil {
+		return nil, err
+	}
+
+	err = sub.Receive(ctx, func(ctx context.Context, pubsubMsg *pubsub.Message) {
+		msg, err := s.unmarshaler.Unmarshal(pubsubMsg)
+		if err != nil {
+			s.logger.Error("could not unmarshal Google Cloud PubSub message", err, nil)
+			return
+		}
+
+		output <- msg
```

Comment:
it should be 

```go
select {
    case output <- msg:
        // ok
     case <-s.closing:
         return errors.New("some error")
}
```
because when client is not consuming and we will call Close it will block forever

Reply:
you also should check that when returning error it will cause nack

Reply:
The callback to `subscription.Receive` doesn't return any error. 

Anyway, when `s.closing` receives, `ctx` is canceled before `output` is closed. So with luck, we should never get a write on a closed `output` channel, because `Receive` is canceled with `ctx`.

If a race happens and `output` is closed before `Receive` is effectively canceled, writing to a closed channel would cause a panic, which may actually make sense here, as long as it is recovered in a sane way. Because it kind of is an abnormal situation.

So if there's no error to return here, maybe I should just recover from panic with an error which causes Nack? Seems a bit artificial though.

Reply:
OK, now I understand what you meant - that there would be no reader for `output` on the other end and it would hang.

2df32f9f8948f714c0bb0705fd609119423a8c5a should fix this, and pass `Ack/Nack` back to the client library, as specified in https://godoc.org/cloud.google.com/go/pubsub#Message.Ack

I'm not entirely sure if I'm using `Nack` correctly. https://godoc.org/cloud.google.com/go/pubsub suggests that sending a `Nack` tells the server "could not process this message, please redeliver ASAP". But if the message is, for example, malformed, it would make no sense to redeliver, rather Ack and put it on a poison queue, I guess?

Reply:
> "could not process this message, please redeliver ASAP"
this is good, when message is malformed poision queue of retry should be used, but it is not responsibility of the PubSub